docs/src/constraints.md

```diff
@@ -1,5 +1,261 @@
+1
+
+julia> con
+1
+```
+However, the reference can be retrieved by querying the model using the symbolic
+name:
+```jldoctest con1
+julia> con = model[:con]
+con : 2 x <= 1.0
+
+julia> con
+con : 2 x <= 1.0
+```
+Because the named variables and constraints are stored in the same namespace,
+creating a constraint with the same name as a variable or an existing constraint
+will result in an error. To overcome this limitation, it is possible to create
+anonymous constraints, just like it is possible to create
+[Anonymous JuMP variables](@ref). This is done by dropping the second argument
+to [`@constraint`](@ref):
+```jldoctest con1
+julia> con = @constraint(model, 2x <= 1)
+2 x <= 1.0
+```
+
+It is also possible use different comparison operators (e.g., `>=` and `==`) to
+create the following types of constraints:
+```jldoctest con1
+julia> @constraint(model, 2x >= 1)
+2 x >= 1.0
+
+julia> @constraint(model, 2x == 1)
+2 x = 1.0
+
+julia> @constraint(model, 1 <= 2x <= 3)
+2 x ∈ [1.0, 3.0]
+```
+
+Note that JuMP normalizes the constraints given by the user by moving all of the
+terms containing variables to the left-hand side, and all of the constant terms
+to the right-hand side. Thus, we get:
+```jldoctest; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, 2x + 1 <= 4x + 4)
+-2 x <= 3.0
+```
+
+## Constraint containers
+
+So far, we've adding constraints one-by-one. However, just like
+[Variable containers](@ref), JuMP provides a mechanism for building groups of
+constraints in *containers*. Three types of constraint containers are supported:
+`Array`s, `JuMPArray`s, and `Dict`ionaries. We explain each of these in the
+following.
+
+### [Arrays](@id constraint_arrays)
+
+The syntax for adding `Array`s of JuMP constraints is very similar to the
+[syntax for creating `Array`s of JuMP variables](@ref Arrays):
+```jldoctest constraint_arrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:3], i * x <= i + 1)
+3-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[1] : x <= 2.0
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+These arrays can be accessed and sliced like normal Julia arrays:
+```jldoctest constraint_arrays
+julia> con[1]
+con[1] : x <= 2.0
+
+julia> con[2:3]
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ con[2] : 2 x <= 3.0
+ con[3] : 3 x <= 4.0
+```
+
+Anonymous containers can also be constructed by dropping the name (e.g. `con`)
+before the square brackets:
+```jldoctest constraint_arrays
+julia> @constraint(model, [i = 1:2], i * x <= i + 1)
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
+ x <= 2.0
+ 2 x <= 3.0
+```
+
+Just like [`@variable`](@ref), JuMP will form an `Array` of constraints when it
+can determine at compile time that the indices are one-based integer ranges.
+Therefore `con[1:b]` will work, but `con[a:b]` will throw an error. If JuMP
```

Comment:
This will only throw an error if you request the container to be `Array`. Otherwise it will return a `JuMPArray`.